File: android/src/main/java/com/zplcode/ZplLibGraphics.java
package com.zplcode;

import android.util.Base64;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.zip.DeflaterOutputStream;

public class ZplLibGraphics {

  public final PixelImage pixels;

  public final int threshold;
  protected Integer x;
  protected Integer y;

  public ZplLibGraphics(PixelImage pixels, int threshold) {
    this.pixels = pixels;
    this.threshold = threshold;
  }

  public void setPoint(Integer x, Integer y) {
    this.x = x;
    this.y = y;
  }

  public String getBodyHeader(boolean insideFormat, int bytesPerRow, int binaryByteCount) {
    StringBuilder result = new StringBuilder();

    if (!insideFormat) result.append("^XA");

    if (x != null && y != null) {
      result.append("^FO").append(x).append(",").append(y);
    }

    result.append("^GFA").append(",");
    result.append(binaryByteCount).append(",");
    result.append(binaryByteCount).append(",");
    result.append(bytesPerRow).append(",");

    return result.toString();
  }

  /**
   *  Reads in a sequence of bytes and prints out its 16 bit
   *  Cylcic Redundancy Check (CRC-CCIIT 0xFFFF).
   *
   *  1 + x + x^5 + x^12 + x^16 is irreducible polynomial.
   *
   */
  public static String getCRCHexString(String bytes)
  {
    int crc = 0x0000;           // initial value
    final int polynomial = 0x1021;    // 0001 0000 0010 0001  (0, 5, 12)
    for (int c = 0; c < bytes.length() ; c++)
    {
      final byte b = (byte) bytes.charAt(c);
      for (int i = 0; i < 8; i++)
      {
        boolean bit = ((b >> (7 - i) & 1) == 1);
        boolean c15 = ((crc >> 15 & 1) == 1);
        crc <<= 1;

        if (c15 ^ bit)
        {
          crc ^= polynomial;
        }
      }
    }

    crc &= 0xffff;
    return Integer.toHexString(crc);
  }


  /**
   * Comprime con el algoritmo zlib que es una abstracción
   * del algoritmo DEFLATE.
   *
   * @param data datos a comprimir
   * @return datos comprimidos
   *
   * @throws java.io.IOException
   */
  public byte[] deflate(byte[] data) throws IOException {
    DeflaterOutputStream deflaterOutputStream = null;
    try {
      final ByteArrayOutputStream compressedImage = new ByteArrayOutputStream(data.length);

      deflaterOutputStream = new DeflaterOutputStream(compressedImage);
      deflaterOutputStream.write(data, 0, data.length);
      deflaterOutputStream.finish();

      return compressedImage.toByteArray();

    } finally {
      if (deflaterOutputStream != null) {
        deflaterOutputStream.close();
      }
    }
  }

  public String getZplCode(boolean insideFormat) throws IOException {
    final int width = pixels.getWidth();
    final int height = pixels.getHeight();
    final byte[] rasterBytes = pixels.getRasterBytes(threshold);

    // LZ77 compression
    final byte[] deflate = deflate(rasterBytes);

    // without compression
    final String z64 = Base64.encodeToString(deflate, Base64.DEFAULT);
    final String crcString = getCRCHexString(z64);

    final int bytesPerRow = rasterBytes.length / height;
    final int binaryByteCount = (width * height) / 8;

    final StringBuilder zpl = new StringBuilder();
    zpl.append(getBodyHeader(insideFormat, bytesPerRow, binaryByteCount));
    zpl.append(":Z64:");
    zpl.append(z64);
    zpl.append(":");
    zpl.append(crcString);
    if (!insideFormat) zpl.append("^XZ");
    return zpl.toString();
  }


}

File: ios/ZplLibGraphics.swift
//
//  ZplLibGraphics.swift
//
//  Created by jbmx on 29/11/22.
//

import Foundation
import Compression

class ZplLibGraphics
{
  /* Imagen a convertir */
  private let pixels: PixelImage
  private let threshold: UInt8
  
  public init(pixels: PixelImage, threshold: UInt8)
  {
    self.pixels = pixels
    self.threshold = threshold
  }
  
  /**
   *  Reads in a sequence of bytes and prints out its 16 bit
   *  Cylcic Redundancy Check (CRC-CCIIT 0xFFFF).
   *
   *  1 + x + x^5 + x^12 + x^16 is irreducible polynomial.
   *
   */
  private static func getCRCHexString(_ bytes: [Character]) -> String
  {
    var crc: Int = 0x0000;           // initial value
    let polynomial: Int = 0x1021;    // 0001 0000 0010 0001  (0, 5, 12)
    
    for c in 0 ..< bytes.count
    {
      let character: Character = bytes[c];
      let b: UInt8 = character.asciiValue!
      
      for i in 0 ..< 8
      {
        let bit: Bool = ((b >> (7 - i) & 1) == 1);
        let c15: Bool = ((crc >> 15 & 1) == 1);
        crc <<= 1;

        if (c15 != bit)
        {
            crc ^= polynomial;
        }
      }
    }

    crc &= 0xffff;
    return String(format:"%2X", crc)
  }
  
  /**
   * Comprime con el algoritmo zlib que es una abstracción
   * del algoritmo DEFLATE.
   *
   * @param data datos a comprimir
   * @return datos comprimidos
   *
   * @throws java.io.IOException
   */
  func deflate(_ data: inout [UInt8]) -> Data
  {
    Utils.log("[ZplLibGraphics]", "deflate: \(data.count)")
    
    let deflater = DeflateStream()
    let (deflated, _) = deflater.write(&data, flush: true)
    return Data(bytes: deflated, count: deflated.count)
    
    // // Crear el búfer de destino
    // let dst_buffer = UnsafeMutablePointer<UInt8>.allocate(
    //  capacity: src_buffer.count)

    // // Comprimir los datos
    // let compressedSize = compression_encode_buffer(
    //  dst_buffer,
    //  src_buffer.count,
    //  &src_buffer,
    //  src_buffer.count,
    //  nil,
    //  COMPRESSION_LZ4  // Seleccione un algoritmo de compresión
    // )

    // return NSData(bytesNoCopy: dst_buffer, length: compressedSize)
  }
  
  /**
   * Obtiene el codigo zpl
   * - Parameter prefixAndSuffix ^XA...^XZ
   * - Returns cadena con el codigo zpl
   */
  public func getZplCode(x: Int, y: Int, prefixAndSuffix: Bool) -> String
  {
    Utils.log("[ZplLibGraphics]", "getZplCode -> x:\(x) y:\(y) prefixAndSuffix:\(prefixAndSuffix)")
    
    let width: Int = pixels.getWidth();
    let height: Int = pixels.getHeight();
    
    var data: [UInt8] = pixels.getRasterBytes(threshold: threshold);
    let bytesPerRow: Int = data.count / height;
    let binaryByteCount: Int = (width * height) / 8;
    
    // LZ77 compression
    //data = dataStatic()
    let deflate: Data = deflate(&data)
    
    // without compression
    let z64: String = (deflate.base64EncodedString())
    let crcString: String = ZplLibGraphics.getCRCHexString([Character](z64));
    
    
    // Label
    var zpl = String()
    
    // Start
    if (prefixAndSuffix) { zpl.append("^XA") }
    
    // Cordenadas
    zpl.append("^FO\(x),\(y)")
    
    zpl.append(
      "^GFA,\(binaryByteCount),\(binaryByteCount),\(bytesPerRow),"
    );
    zpl.append(":Z64:");
    zpl.append(z64);
    zpl.append(":");
    zpl.append(crcString);
    
    // End
    if (prefixAndSuffix) { zpl.append("^XZ"); }
    
    return zpl + "\n"
  }
  
}
